charmcraft/models/project.py

```diff
@@ -541,24 +492,28 @@ def from_yaml_file(cls, path: pathlib.Path) -> Self:
 
         return project
 
-    @pydantic.root_validator(pre=True, allow_reuse=True)
-    def preprocess(cls, values: dict[str, Any]) -> dict[str, Any]:
+    @pydantic.model_validator(mode="before")
+    @classmethod
+    def _preprocess(cls, values: dict[str, Any]) -> dict[str, Any]:
         """Preprocess any values that charmcraft infers, before attribute validation."""
         if "type" not in values:
             raise ValueError("Project type must be declared in charmcraft.yaml.")
 
         return values
 
-    @pydantic.validator("parts", pre=True, always=True, allow_reuse=True)
```

Comment:
> `always=True`

so this needs to be handled on the field now right? didn't this change break anything?

Reply:
The validation is still handled like the upstream one, but unfortunately this preprocessing can't work on the field itself :-/